### DISCUSSION TITLE
ESP32 multiple output setup

### DISCUSSION THREAD

**Discussion by martin072** [arendst/Tasmota#14782] [Q&A] [answered]
I am having trouble setting up an ESP32 with 3 outputs:

1  Output for WS2812 
1 Output for PWM
1 Output for a regular relay. 

However, on the web UI I only see 2 toggle's. See screen shots for the setup. 

Edit:
So it seems that PWM and a WS2812 don't go together? It looks like there are 2 sliders for dimming however the both show the same when issuing a dimmer command.. 

Is it by design PWM and WS2812 won't work together?

<img width="455" alt="Screenshot 2022-02-08 at 22 19 07" src="https://user-images.githubusercontent.com/24524506/153077731-422333da-54d7-4f16-954a-3c13617029cc.png">
<img width="421" alt="Screenshot 2022-02-08 at 22 18 56" src="https://user-images.githubusercontent.com/24524506/153077741-9188c105-c92d-4d18-9388-5b56f2744dc1.png">

**Comment by sfromis** [comment, marked as answer]
You guessed it. Tasmota can only have one type of light, and default setup for both PWM and Neopixel is as a light. When trying both, the Neopixel setup "wins", making the PWM light inoperative. 

But that's not the end of the story, with most recent dev ESP32 Tasmota, you can still use a non-light PWM if you configure it as PWM6, and use the PWM6 command to select duty cycle up to 1023. This provides no automatic slider or toggle in the web UI. By default, PWM1-5 is RGBCW for a light, but the next 11 channels are for any PWM purpose.

Without the 16-channel PWM support with most recent dev ESP32 Tasmota (or on ESP8266), you can stay on PWM1, and use `SetOption15` to select basic PWM control instead of lights, and the PWM1 command for that.

**Reply by martin072** [reply]
Thanks for the quick answer!

When removing the PWM, so only using the WS2812, I still see two dimmer sliders? that seem to share the same value. Any idea why this is? I don't see this on an ESP8266 with a WS2812 strip. 

When I remove the relay output (so the WS2812 is the only output) I only have one dimmer slider?

**Reply by sfromis** [reply]
Did not see the extra slider from your initial post when doing the bad config of PWM1 + WS2812 on ESP32 with Tasmota 10.1.0.8. Removing the PWM from such an unusable config leaves me with the usual 3 sliders for WS2812.

You may have been messing around with other config options trying to make your unusable config work somehow. I'd revert to initial settings, only configure WS2812, and after verifying that, maybe add PWM6.

**Reply by martin072** [reply]
Already did a reset to firmware defaults. same result. Using 10.1.0 release

Decided to stick with WS2812 and a relay only. Will use the PWM on a separate device.

see pics below. When removing the relay from the config it shows correct. 
<img width="384" alt="Screenshot 2022-02-09 at 10 28 43" src="https://user-images.githubusercontent.com/24524506/153166189-c4d87005-d7da-4451-82f8-5abe6395dbfb.png">
<img width="401" alt="Screenshot 2022-02-09 at 10 28 28" src="https://user-images.githubusercontent.com/24524506/153166202-03e8fbd2-5934-460d-9be6-0b0a96899cb0.png">

Relay removed:

<img width="437" alt="Screenshot 2022-02-09 at 10 31 44" src="https://user-images.githubusercontent.com/24524506/153166539-66656f15-6c1b-4eb4-8202-2e9c4554c2ba.png">
<img width="391" alt="Screenshot 2022-02-09 at 10 31 25" src="https://user-images.githubusercontent.com/24524506/153166550-71b09bb4-10a4-41ac-8410-8a1c1cbe0262.png">

**Reply by sfromis** [reply]
My test config for PWM+WS2812 actually included gpios for both relay and button, very similar to your setup, except for I2C pins. Adding those, I still do not get the extra slider. What Tasmota version do you use? I'd suggest most recent dev. 

Your images suggest you doing additional config instead of **only** adding WS2812 as first step after "factory reset". For testing, it generally works best to add *as little as possible* at a time, instead of jumping to the desired config including other settings.

**Reply by martin072** [reply]
Yep did that. Reset, then first the WS2812, all looks fine. The added the button, and then the double sliders appear. Didn't do any other config other than defining the outputs via the GUI. 

Will try latest dev. release to see if it makes a difference. 

thanks

**Reply by sfromis** [reply]
What does "defining the outputs" mean? Including setting WebButton? Sounds different from "as little as possible".....

**Reply by martin072** [reply]
The webbutton command is the only one given in the console, rest of the setup is done via the gui, only the outputs nothing else, no mqtt etc 

Would be surprised if a webbutton command causes this ;)

**Reply by sfromis** [reply]
Not claiming that WebButton is the cause, just that it was unclear how much more than *as little as possible* you were doing. And my basic setup does not have the extra slider when defining WS2812+PWM+Button+Relay.

**Reply by martin072** [reply]
Still the same, got developer release. reflashed. first a relay, shows one output, then restart and the ws2812, shows two outputs and double slider. 

No idea what's causing this. will erase the chip and flash it from scratch.

**Reply by martin072** [reply]
Erased flash with Esptool, refreshed it, still the same 🧐

**Reply by pinoyyid** [reply]
I've seen the same behaviour, ie. adding an unsupported output light and then removing it does not restore the GUI controls to their original state. It took several resets and reboots before I saw the expected behaviour.
